src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -0,0 +1,335 @@
+#include <protocols/bdx/BdxMessages.h>
+#include <protocols/bdx/BdxTransferSession.h>
+
+#include <nlunit-test.h>
+
+#include <support/CodeUtils.h>
+#include <support/UnitTestRegistration.h>
+
+using namespace ::chip;
+using namespace ::chip::BDX;
+
+void SendAndVerifyQuery(nlTestSuite * inSuite, void * inContext, TransferSession & sender, TransferSession & receiver,
+                        TransferSession::OutputEvent & outEvent)
+{
+    // Verify that receiver emits BlockQuery message
+    CHIP_ERROR err = receiver.PrepareBlockQuery();
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    receiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle queryMsgToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !queryMsgToSend.IsNull());
+
+    // Pass BlockQuery to sender
+    err = sender.HandleMessageReceived(queryMsgToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Verify sender emits QueryReceived event
+    sender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_QueryReceived);
+}
+
+void SendAndVerifyArbitraryBlock(nlTestSuite * inSuite, void * inContext, TransferSession & sender, TransferSession & receiver,
+                                 TransferSession::OutputEvent & outEvent, bool isEof)
+{
+    CHIP_ERROR err                         = CHIP_NO_ERROR;
+    uint16_t maxBlockSize                  = sender.GetTransferBlockSize();
+    System::PacketBufferHandle fakeDataBuf = System::PacketBuffer::NewWithAvailableSize(maxBlockSize);
+    if (fakeDataBuf.IsNull())
+    {
+        NL_TEST_ASSERT(inSuite, false);
+        return;
+    }
+
+    uint8_t * fakeBlockData = fakeDataBuf->Start();
+
+    // Provide Block data and verify sender emits Block message
+    TransferSession::BlockData blockData;
+    blockData.Data   = fakeBlockData;
+    blockData.Length = maxBlockSize;
+    blockData.IsEof  = isEof;
+
+    err = sender.PrepareBlock(blockData);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    sender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle blockMsgToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !blockMsgToSend.IsNull());
+
+    // Pass Block message to receiver and verify Block is received
+    err = receiver.HandleMessageReceived(blockMsgToSend.Retain());
```

Comment:
Move, or better just assert `!outEvent.MsgData.IsNull()` above and `HandleMessageReceived(outEvent.MsgData.Retain())` here.